cmd/fs-v1-metadata.go

```diff
@@ -237,58 +250,117 @@ func newFSMetaV1() (fsMeta fsMetaV1) {
+		fsFormatV1,
+		fsFormatV2,
+		// New supported versions here.
+	}
+
+	// Check for supported format versions.
+	for _, version := range supportedFormatVersions {
+		if formatCfg.FS.Version == version {
+			return true
 		}
-		return nil, err
 	}
-	defer rlk.Close()
+	return false
+}
 
-	formatBytes, err := ioutil.ReadAll(rlk)
-	if err != nil {
-		return nil, err
+// errFSFormatOld- old fs format.
+var errFSFormatOld = errors.New("old FS format found")
+
+// Checks if the loaded `format.json` is valid and
+// is expected to be of the requested version.
+func checkFormatFS(format *formatConfigV1, formatVersion string) error {
+	if format == nil {
+		return errUnexpected
 	}
 
-	format := &formatConfigV1{}
-	if err = json.Unmarshal(formatBytes, format); err != nil {
-		return nil, err
+	// Validate if we have the same format.
+	if format.Format != "fs" {
+		return fmt.Errorf("Unable to recognize backend format, Disk is not in FS format. %s", format.Format)
 	}
 
-	return format, nil
-}
+	// Check if format is currently supported.
+	if !isFSValidFormat(format) {
+		return errCorruptedFormat
+	}
 
-// writes FS format (format.json) into minioMetaBucket.
-func saveFormatFS(formatPath string, fsFormat *formatConfigV1) error {
-	metadataBytes, err := json.Marshal(fsFormat)
-	if err != nil {
-		return err
+	// Check for format version is current.
+	if format.FS.Version != formatVersion {
+		return errFSFormatOld
 	}
 
+	return nil
+}
+
+// Initializes a new `format.json` if not present, validates `format.json` if already present.
+func initFormatFS(fsPath, fsUUID string) error {
+	fsFormatPath := pathJoin(fsPath, minioMetaBucket, fsFormatJSONFile)
+
 	// fsFormatJSONFile - format.json file stored in minioMetaBucket(.minio.sys) directory.
-	lk, err := lock.LockedOpenFile(preparePath(formatPath), os.O_CREATE|os.O_WRONLY, 0600)
+	lk, err := lock.LockedOpenFile(preparePath(fsFormatPath), os.O_RDWR|os.O_CREATE, 0600)
 	if err != nil {
-		return err
+		return traceError(err)
 	}
 	defer lk.Close()
 
-	_, err = lk.Write(metadataBytes)
+	var format = &formatConfigV1{}
+	_, err = format.ReadFrom(lk)
+	// For all unexpected errors, we return.
+	if err != nil && errorCause(err) != io.EOF {
+		return fmt.Errorf("Unable to load 'format.json', %s", err)
+	}
+
+	// If we couldn't read anything, we assume a default
+	// (empty) upload info.
+	if errorCause(err) == io.EOF {
+		format = newFSFormatV2()
+	}
+
+	// Validate loaded `format.json`.
+	err = checkFormatFS(format, fsFormatVersion)
+	if err != nil && err != errFSFormatOld {
+		return fmt.Errorf("Unable to validate 'format.json', %s", err)
+	}
+
+	// Disk is in old format migrate object metadata.
+	if err == errFSFormatOld {
+		if err = migrateFSObject(fsPath, fsUUID); err != nil {
+			return err
+		}
+
+		// Initialize format v2.
+		format = newFSFormatV2()
+
+		// Initialize format.json.
+		if _, err = format.WriteTo(lk); err != nil {
+			return fmt.Errorf("Unable to initialize 'format.json', %s", err)
+		}
+	}
```

Comment:
still a problem, because on a fresh export directory we are not writing format to disk, i.e we endup with empty file created at line 322